### PR TITLE
Feat/schema management v2

### PR DESCRIPTION
**Pull request by loicgreffier**

### CODE REVIEW

File: api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.michelin.ns4kafka.services.schema.client.entities;
+
+import lombok.*;
+
+@Getter
+@Builder
+public class SchemaCompatibilityRequest {
+    private String compatibility;
+}
```
Comment (twobeeb) [inline api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java:9]: Why do we need this class ? Micronaut @Body String enough ?

File: api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityResponse.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Compatibility level access method
+     *
+     * @return compatibilityLevel property
+     */
+    @JsonProperty("compatibilityLevel")
+    public Schema.Compatibility compatibilityLevel() {
+        return compatibilityLevel;
+    }
+}
```
Comment (twobeeb) [inline api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityResponse.java:34]: Why do we need this class ? Micronaut @Body String enough ?

File: cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java [hunk 3]
```diff
@@ -71,14 +73,25 @@ public Resource getSingleResourceWithType(ApiResource apiResource, String namesp
 
     public HttpResponse<Resource> apply(ApiResource apiResource, String namespace, Resource resource, boolean dryRun) {
         try {
+            // If the resource is a subject, load the schema
+            if (resource.getKind().equals("Schema")) {
+                fileService.loadSchemaFileContent(resource);
```
Comment (twobeeb) [inline cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java:78]: loadSchema only if ``schemaFile`` is there. Remember 
````yaml
schema: {"type": "string"}
# or
schema: |
    {
      "type": "string"
    }
# or
schemaFile: schema.avsc
````

File: cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java [hunk 4]
```diff
@@ -71,14 +73,25 @@ public Resource getSingleResourceWithType(ApiResource apiResource, String namesp
 
     public HttpResponse<Resource> apply(ApiResource apiResource, String namespace, Resource resource, boolean dryRun) {
         try {
+            // If the resource is a subject, load the schema
+            if (resource.getKind().equals("Schema")) {
```
Comment (twobeeb) [inline cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java:77]: Maybe this shouldn't be in apply. We should discuss this.

**Review by twobeeb — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java (hunk 1 above)
Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java:9]: The Schema Registry requires a JSON object as body : https://docs.confluent.io/platform/current/schema-registry/develop/api.html#put--config-(string-%20subject) with a "compatibility" field :

{
  "compatibility": "FULL"
}

A simple map (e.g., <"compatibility", "FULL">) can be used as body, in place of the custom SchemaCompatibilityRequest object.

File: api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        SchemaCompatibilityState state = this
+                .schemaService.updateSubjectCompatibility(ns, existingSchemaOptional.get(), compatibility);
+
+        sendEventLog("SchemaCompatibilityState",
+                existingSchemaOptional.get().getMetadata(),
+                ApplyStatus.changed,
+                existingSchemaOptional.get().getSpec().getCompatibility(),
+                compatibility);
```
[the remark this answers is not in the document]
Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java:185]: Fixed

File: api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        if (dryrun) {
+            // Cannot compute the apply status before the registration
+            return HttpResponse.ok(schema);
+        }
+
+        Optional<Schema> existingSchemaOptional = this.schemaService.getLatestSubject(ns, schema.getMetadata().getName());
+        Integer registeredSchemaId = this.schemaService.register(ns, schema);
+
+        ApplyStatus status = existingSchemaOptional.isEmpty() ? ApplyStatus.created : registeredSchemaId > existingSchemaOptional
+                .get().getSpec().getId() ? ApplyStatus.changed : ApplyStatus.unchanged;
```
[the remark this answers is not in the document]
Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java]: Fixed

File: api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    /**
+     * Get all the schemas by namespace
+     *
+     * @param namespace The namespace
+     * @return A list of schemas
+     */
+    @Get
+    public List<Schema> list(String namespace) {
+        Namespace ns = getNamespace(namespace);
+        return this.schemaService.findAllForNamespace(ns);
```
[the remark this answers is not in the document]
Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java]: Fixed

File: api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java (hunk 1 above)
Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/services/schema/client/entities/SchemaCompatibilityRequest.java]: Fixed, `SchemaCompatibilityRequest` replaced by single body String

File: cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java (hunk 3 above)
Comment (loicgreffier) [reply cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java]: Fixed

**Pull request merged**